### PR TITLE
application to generate scales for bump3d

### LINKED ISSUE #709: Horizontal and vertical scales for the 3D bump correlation operator
## Description
There are currently a few bottlenecks in the current default static B (SocaError) which was designed to shorten the initialization time at the detriment of the computing time of a B multiply operation. 

## Solution
Revisit the use of 3D bump, this will result in a faster B multiply at the expense of a slower initialization time (done only once).

## Work to do
- Develop an application that will generate horizontal and vertical scales similar to what is currently being used in SocaError
- Add relevant ctests:
   - generate the files containing the scales needed as input to bump
   - initialize a bump using above scales
   - test using the dirac application for example

### PR DESCRIPTION
**Pull request by guillaumevernieres**
## Description
See #709 

#### What was done
Addition of an application that writes (parametric) decorrelation length scales to files. The output can be read by bump.

Here's an example `yaml` that can be passed to ``soca_setcorscales.x``:

```yaml
resolution:
  mom6_input_nml: ./inputnml/input.nml
  fields metadata: ./fields_metadata.yml

date: 2018-04-15T00:00:00Z

corr variables: [socn, tocn, ssh, hocn, cicen, hicen]

scales:
  vert layers: 100 # in units of layer
  socn:
    rossby mult: 40.0
    min grid mult: 5.0
  tocn:
    rossby mult: 30.0
    min grid mult: 6.0
  hocn:
    rossby mult: 40.0
    min grid mult: 10.0
  ssh:
    rossby mult: 20.0
    min grid mult: 7.0
  cicen:
    rossby mult: 0.0
    min grid mult: 8.0
    min: 1000000
  hicen:
    rossby mult: 0.0
    min grid mult: 9.0
    min: 5000000

rh output:
  datadir: Data
  exp: cor_rh
  type: incr

rv output:
  datadir: Data
  exp: cor_rv
  type: incr
```

#### How was it tested
3 ctests:
-  example use of ``soca_setcorscales.x`` 
- initialization of bump using the (unrealistic) scales specified by the files created by ``soca_setcorscales.x``
- dirac test

fixes #709

### CODE REVIEW

**Review by kbhargava — APPROVED** [review]
Builds and ctests pass on orion except the following usual ones that fail on orion in comparison step.
```
	  5 - test_soca_convertstate (Failed)
	 44 - test_soca_dirac_soca_cor_nicas_scales (Failed)
	 48 - test_soca_dirac_soca_mask (Failed)
	 53 - test_soca_enshofx (Failed)
	 55 - test_soca_3dvarbump (Failed)
```

**Review by benjaminmenetrier — APPROVED** [review]
Just one comment: `gau2gc` will disappear with #703. Could you replace it with `3.57_kind_real`? It will make the future merge easier. Thanks!

**Review by clouden90 — APPROVED** [review]
Codes looks good. Tested on Orion. test_soca_dirac_soca_cor_nicas_scales failed due to a 0.000001 difference compared to reference.

**Comment by guillaumevernieres** [comment]
> Just one comment: `gau2gc` will disappear with #703. Could you replace it with `3.57_kind_real`? It will make the future merge easier. Thanks!

~Thanks for the heads-up @benjaminmenetrier . We'll make the change in a dedicated PR since it is used in other places.~
Never mind, making the change in that PR.

File: src/soca/Increment/soca_increment_mod.F90 [hunk 1]
```diff
@@ -513,6 +524,52 @@ subroutine soca_increment_change_resol(self, rhs)
...
+subroutine soca_vert_scales(self, vert)
+  class(soca_increment), intent(inout) :: self
+  real(kind=kind_real),  intent(in)    :: vert
+
+  integer :: i, jz
+
+  ! compute scales
+  do i=1,size(self%fields)
+    do jz=1,self%fields(i)%nz
+      self%fields(i)%val(:,:,jz) = gau2gc*gau2gc*self%geom%mask2d(:,:)*vert
```
Comment (guillaumevernieres) [inline src/soca/Increment/soca_increment_mod.F90]: ha ha ...

Comment (benjaminmenetrier) [reply src/soca/Increment/soca_increment_mod.F90]: Always a good practice pushing as many corrections as possible in the same PR!

Comment (guillaumevernieres) [reply src/soca/Increment/soca_increment_mod.F90]: I was testing the reviewers of course.

**Pull request merged**